sqsworkers/crew.py

```diff
@@ -1,327 +1,200 @@
+
+                    self.logger.info(
+                        "length of messages ({}) below minimum set to pass to message processor ({}). polling".format(
+                            len(messages), self.minimum_messages
+                        )
+                    )
+
+                    messages += self.queue.receive_messages(
+                        AttributeNames=["All"],
+                        MessageAttributeNames=["All"],
+                        MaxNumberOfMessages=self.max_number_of_messages,
+                        WaitTimeSeconds=self.wait_time,
+                    )
+
             else:
-                self.crew.logger.warn('worker %s is dead, hiring a new one...' % (worker.worker_name))
-                self.crew.statsd.increment('workers.dead', 1, tags=[self.crew.name])
-                new_worker = Worker(self.crew)
-                self.crew.logger.info('hired a new worker %s, staring it up...' % (new_worker.worker_name))
-                new_worker.start()
-                self.crew.logger.info('started a new worker %s' % (new_worker.worker_name))
-                good_workers.append(new_worker)
-        self.crew.logger.info('total number of workers after supervision: %s' % (str(len(good_workers))))
-        self.crew.statsd.gauge('workers.employed', len(good_workers), tags=[self.crew.name])
-        self.crew.workers = good_workers
+
+                messages = self.queue.receive_messages(
+                    AttributeNames=["All"],
+                    MessageAttributeNames=["All"],
+                    MaxNumberOfMessages=self.max_number_of_messages,
+                    WaitTimeSeconds=self.wait_time,
+                )
+
+            if messages:
+
+                metadata = [MessageMetadata(m) for m in messages]
+
+                self.logger.info(
+                    f"processing the following {len(messages)} messages in bulk: {metadata}"
+                )
+
+                task: futures.Future = self._executor.submit(
+                    self.message_processor(messages).start
+                )
+
+                task.add_done_callback(
+                    partial(
+                        self._task_complete,
+                        messages=messages,
+                        metadata=metadata,
+                    )
+                )
+
+                self.statsd.increment(
+                    "process.record.start", len(messages), tags=[]
+                )
+
+            time.sleep(self.polling_interval)
+
+    def _task_complete(
+        self,
+        f: futures.Future,
+        messages: Iterable[Any],
+        metadata=Optional[List[MessageMetadata]],
+    ):
+        """Clean up after task and do any necessary logging."""
+
+        messages = list(messages)
+
+        metadata: List[MessageMetadata] = [
+            MessageMetadata(m) for m in messages
+        ] if metadata is None else metadata
+
+        exception = f.exception()
+
+        if exception is not None:
```

Comment:
What if only one message fails from the bulk messages that you picked up? Since you are giving the messagesprocessor.start(messages) when it returns would it not return not all messages? It seems like you are handling as if all messages failed, but what if only one does?

Reply:
Yep, it's all or nothing if there's a group of messages. I could implement logic to try and parse out only those messages that failed, but I feel like the current implementation is a decent enough tradeoff to avoid that complexity. Otherwise it's almost like going back to handling only one message at a time; no reason to have a bulk processor.

Reply:
mmkay new changes should be reflected that alter bulk behavior